kge/indexing.py

```diff
@@ -2,6 +2,112 @@
 from collections import defaultdict, OrderedDict
 import numba
 import numpy as np
+from typing import Dict, List, Iterator, Tuple
+
+
+class KvsAllIndexDict:
+    def __init__(self, triples: torch.Tensor, key_cols: List, value_column: int,
+                 default_factory: type):
+        """
+        Construct KvsAllIndexDict
+        Args:
+            triples: data
+            key_cols: column indicators of data, giving the keys of the dictionary
+            value_column: column indicator of data, giving the values of the dictionary
+            default_factory: default return type
+        """
+        self.key_cols = key_cols
+        self.value_column = value_column
+        self.data_sorted = self.sort_data_by_keys(triples, key_cols, value_column)
+        self.index, self.offset = self.construct_index()
+
+        self.default_factory = default_factory
+
+    def __getitem__(self, key, default_return_value=None):
+        try:
+            key_range = self.index[key]
+            return self.data_sorted[key_range[0]:key_range[1], self.value_column]
+        except KeyError:
+            if default_return_value is None:
+                return self.default_factory()
+            return default_return_value
+
+    def __len__(self):
+        return len(self.index)
+
+    def get(self, key, default_return_value=None):
+        return self.__getitem__(key, default_return_value)
+
+    def keys(self):
+        return self.index.keys()
+
+    def values(self):
+        values = []
+        for value in self.index.values():
+            values.append(self.data_sorted[value[0]:value[1], self.value_column])
+        return values
+
+    def items(self) -> Iterator[Tuple[Tuple[int, int], torch.Tensor]]:
+        keys = self.keys()
+        values = self.values()
+        return zip(keys, values)
```

Comment:
Do we every use this? It's very inefficient...

Reply:
it is indeed inefficient, but it is only used in test_dataset, since we expect working on dictionaries